rfcs/plugins.md

```diff
@@ -41,223 +42,140 @@ Conflicts between plugins should be easy to determine from their
 called, those functions are for server plugins.
 
-### Config
 
-The `config` plugin property specifies which config options will be
-forwarded from the config file section for the plugin. The value
-should be a
-[Joi validator](https://github.com/hapijs/joi/blob/master/API.md) to
-validates the config options.
+### Activate
 
-So a `config` key like:
+The `activate` plugin property must be present and contain a function
+that accepts an object with the following properties:
+  - `config`: config options from the config.toml section for the
+    plugin. This includes user overrides which are not specified in
+    the `config` plugin property.
+  - `metadata`: the 
+  - `done`: a callback to be called with the result when the plugin is
+    ready.
 
-```js
-config: {
-   someKey: Joi.boolean(),
-   some_other_key: Joi.string(),
-   another: Joi.string().default('foo'),
-}
-```
+The `done` function should be called with an object:
 
-Will allow the following section in `config.toml`:
+The function should return an object which describes how the plugin
+wishes to modify the Horizon server.  This object will be passed to
+`deactivate` later.  Recognized fields are:
+  - `httpRoute`: specifies the http route that will be owned by the plugin [Http route](#http-route)
+  - `requests`: specified new request types to accept over the websocket connection [requests](#requests)
 
-```toml
-[plugin.my-plugin-name]
-someKey = true
-some_other_key = 'some kinda string'
-```
 
-The values from this config section will be passed to both the
-`processConfig` function for commands, as well as the
-[`activate`](#activate) function for server plugins. The object passed
-will look like:
+#### HTTP route
 
-```js
-{ someKey: true,
-  some_other_key: 'some_kinda_string',
-  another: 'foo',
-}
-```
+The `httpRoute` property is an object whose fields define
+the http routes the plugin will receive requests from. The
+values are request handler functions for each route. The functions
+receive three arguments:
+ - A `request` object
+ - A `response` object
 
-### HTTP route
-
-The `httpRoute` plugin property is an object with a single key
-defining the http route the plugin will receive requests from. The
-value is the request handler for the route. The function receives
-three arguments:
- - An Express `request` object
- - An Express `response` object
- - A state object of whatever kind is returned by the plugin's
-   [activate](#activate) function
+The route field name is appended to the plugin name for all routes.
+This field must be either an empty string or begin with a `/`, e.g.
+`""`, `"/foo"` to avoid conflicting with other plugins.
 
 The handler should deal with whatever HTTP methods come to the route.
 
-Example:
+Example for a plugin named `graphql`:
 
 ```js
 httpRoute: {
-    'hello-world': (req, res, state) => {
-        res.send(`Hello ${state.value}`);
-            res.end()
-    },
+  '': (req, res) => {
```

Comment:
I believe the idea is that the function is passed all HTTP requests except upgrade, and it is up to the plugin to branch based on the request type.